cbits/shared.c

```diff
@@ -0,0 +1,487 @@
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "shared.h"
+#include "sha1.h"
+
+// #define FSHOOK_SHARED_DEBUG
+
+#ifdef SHARED_TEST
+#define FSHOOK_SHARED_DEBUG
+#endif
+
+#ifdef FSHOOK_SHARED_DEBUG
+#define LOG(fmt, params...) do { \
+    fprintf(stderr, "fshooks-shared<pid=%d>: " \
+            fmt "\n", getpid(), ##params); \
+    fflush(stderr); \
+} while (0)
+#else
+#define LOG(fmt, params...) do { if (0) { printf(fmt "\n", ##params); } } while (0)
+#endif
+
+#define KEY_DATA_SIZE                 20
+#define MAX_DISTANCE_FOR_RESIZE       0x10
+#define INITIAL_TABLE_ORDER           9
+
+typedef struct _key_hash {
+    union {
+        unsigned char data[KEY_DATA_SIZE];
+        struct {
+            uint64_t a;
+            uint64_t b;
+            uint32_t c;
+        };
+    };
+    uint32_t blob_offset;
+} key_hash;
+
+typedef struct _shmem_header {
+    size_t table_start;
+    size_t table_order;
+    size_t table_end;
+    size_t items_used;
+    size_t next_blob_offset;
+    size_t mappable_size;
+} shmem_header;
+
+typedef struct _shmem_context {
+    int fd;
+    int ro_fd;
+    size_t size;
+    size_t file_size;
+    size_t mapped_size;
+    pthread_mutex_t mutex;
+    void *root;
+} shmem_context;
+
+static void string_to_key_hash(const char *str, size_t len, key_hash *out)
+{
+    sha1(out->data, (unsigned char *)str, len);
+}
+
+static bool key_hash_is_empty(const key_hash *existing_key)
+{
+    return (existing_key->a == 0  &&
+            existing_key->b == 0  &&
+            existing_key->c == 0);
+}
+
+static uint64_t key_hash_to_full_idx(const key_hash *key)
+{
+    return key->a ^ key->b ^ key->c;
+}
+
+static void shmem_remap(shmem_context *shmem_ctx)
+{
+    const size_t mapped_size = (shmem_ctx->size + 0xfff) & ~0xfff;
+    const bool is_readonly = shmem_ctx->fd == -1;
+
+    if (shmem_ctx->mapped_size != mapped_size) {
+        LOG("remap to size 0x%lx", mapped_size);
+
+        if (shmem_ctx->mapped_size) {
+            munmap(shmem_ctx->root, shmem_ctx->mapped_size);
+        }
+
+        const int fd = is_readonly ? shmem_ctx->ro_fd : shmem_ctx->fd;
+        shmem_ctx->root = mmap(NULL, mapped_size,
+                               PROT_READ | (is_readonly ? 0 : PROT_WRITE),
+                               MAP_SHARED,
+                               fd,
+                               0);
+        if (shmem_ctx->root == MAP_FAILED) {
+            if (errno == EBADF) {
+                /*
+                 * The process may have did a for-loop to
+                 * close all FDs on shutdown.
+                 */
+                return;
+            }
+        }
+
+        shmem_ctx->mapped_size = mapped_size;
+
+        if (!is_readonly) {
+            shmem_header *header = ((shmem_header *)shmem_ctx->root);
+            header->mappable_size = mapped_size;
+        }
+
+        assert(shmem_ctx->root != MAP_FAILED);
+    }
+}
+
+static void shmem_resize(shmem_context *shmem_ctx)
```

Comment:
I don't understand this. Shouldn't we just ftruncate just before each mmap call if file_size is less than the new mapped size? ...either that or I'm missing something